src/org/javarosa/xpath/expr/XPathFuncExpr.java

```diff
@@ -356,6 +356,26 @@ public Object eval (DataInstance model, EvaluationContext evalContext) {
             int pos = str.indexOf(substr);
             // XPath reference states that we should return the empty string when we don't find the substring
             return pos >= 0 ? str.substring(pos + substr.length()) : "";
+        } else if (name.equals("translate") && args.length == 3) {
+            String str = toString(argVals[0]);
+            String fromChars = toString(argVals[1]);
+            String toChars = toString(argVals[2]);
+            StringBuilder result = new StringBuilder();
+            // iterate thru each char in the original string
+            for (int i = 0; i < str.length(); i++) {
+                char from = str.charAt(i);
+                // determine if there is a mapping for this char
+                int fromPos = fromChars.indexOf(from);
+                if (fromPos == -1) {
+                    // no mapping for this char, so just add it to result unchanged
+                    result.append(from);
+                } else if (fromPos < toChars.length()) {
```

Comment:
`toChars.length()` could be pre-computed before the `for` block to avoid computing it each loop iteration.

Reply:
Done. Good catch. Thnx.

BTW I left the other potential candidate for this optimization - namely the for loop terminating expression - unchanged, for internal consistency with existing toNumeric().